adding new CI golang tutorial (#1591)
harness/developer-hub: tutorials/ci-pipelines/build/ci-golang.md

---
sidebar_position: 5
title: Build and test a Go app
description: Use a CI pipeline to build and test a Go application.
keywords: [Hosted Build, Continuous Integration, Hosted, CI Tutorial]
slug: /ci-pipelines/build/golang
---

# Build and test a Go app

> Build and test a [Go](https://go.dev/) application using Harness CI's Linux Cloud infrastructure.

Reviewer: ```suggestion
Build and test a [Go](https://go.dev/) application using a Linux platform on [Harness Cloud build infrastructure](/docs/continuous-integration/use-ci/set-up-build-infrastructure/use-harness-cloud-build-infrastructure).
```

                          - report.xml
          platform:
            os: Linux
            arch: Amd64
          runtime:
            type: Cloud
            spec: {}
  properties:
    ci:
      codebase:
        connectorRef: [your-code-repo-connector-ID] # replace with your connector ID
        repoName: [your-repository-name] # replace with your repository name
        build: <+input>
```

</details>

<details>
<summary>Use multiple versions of Go</summary>

```yaml
pipeline:
  name: Build and test Go app
  identifier: Build_and_test_Go_app
  projectIdentifier: default
  orgIdentifier: default
  stages:
    - stage:
        name: Build
        identifier: Build
        type: CI
        strategy:
          matrix:
            goVersion:
              - "1.19"
              - "1.20"
        spec:
          caching:
            enabled: true
          cloneCodebase: true
          execution:
            steps:
              - step:
                  type: Run
                  identifier: installgo
                  name: Install Go
                  spec:
                    shell: Sh
                    command: |-
                      export GOPATH=$HOME/go

>                       go install golang.org/dl/go<+matrix.goVersion>@latest

Reviewer 2: Our VM image comes with a [few versions of Golang](https://github.com/wings-software/virtual-environments/blob/main/images/linux/Ubuntu2204-Readme.md#go) pre-installed. Can we mention how to switch between these already installed versions without having to install them again (unless the install is a no-op for already installed versions)? Are we no longer advocating using GH Actions for setup-<language> steps?

cc: @rustd @bradrydzewski

Reviewer: For switching versions: I think that Jim included an example of switching versions - @jimsheldon ?

Reviewer 3: agreed, I would like to see us optimize the tutorials to take advantage of pre-installed software instead of asking our users to install software.  @hemanthmantri can you suggest the correct pipeline configuration to do a go build / test that uses the pre-installed Go version?

Author: @hemanthmantri there is a link to documentation under "Setup Go" that describes the OS and tools installed.

The versions of Go are installed under `/opt/hostedtoolcache/go`, I need to work with someone who can help me recommend the best method if switching between these versions.

                      $GOPATH/bin/go<+matrix.goVersion> download
              - step:
                  type: Run
                  identifier: build
                  name: Build
                  spec:
                    shell: Sh
                    command: |-
                      export GOPATH=$HOME/go
                      $GOPATH/bin/go<+matrix.goVersion> build
              - step:
                  type: Run
                  name: Test
                  identifier: test
                  spec:
                    shell: Sh
                    command: |-
                      export GOPATH=$HOME/go
                      $GOPATH/bin/go<+matrix.goVersion> install github.com/jstemmer/go-junit-report/v2@latest

>                       $GOPATH/bin/go<+matrix.goVersion> test -v ./... | tee report.out

Reviewer 2: Since matrix steps run in parallel, they will all write to the same `report.out`, right? Use `report_<matrix.goVersion>.out`

Reviewer 3: similar to above comment, let's optimize the matrix to take advantage of the existing Go versions on the machine

Author: Thanks for catching this, I pushed a fix.

                  type: Run
                  identifier: build
                  name: Build
                  spec:
                    shell: Sh
                    command: |-
                      go build
              - step:
                  type: Run
                  identifier: test
                  name: Test
                  spec:
                    shell: Sh
                    command: |-
                      go test -v ./...
```

```mdx-code-block
</TabItem>

<TabItem value="Self-hosted">
```

```yaml
              - step:
                  type: Run
                  identifier: build
                  name: Build
                  spec:
                    connectorRef: [your-docker-connector-ID] # replace with your connector ID
                    image: golang
                    command: |-
                      go build
              - step:
                  type: Run
                  identifier: test
                  name: Test
                  spec:
                    connectorRef: [your-docker-connector-ID] # replace with your connector ID
                    image: golang
                    command: |-
                      go test -v ./...
```

> 

Author: I think we will hold off on this for now, just trying to keep this minimal if possible

                    image: golang
                    command: |-
                      go get example.com/my-go-module
```

```mdx-code-block
</TabItem>
</Tabs>
```

## Cache dependencies

Add caching to your stage.

```mdx-code-block
<Tabs>
<TabItem value="Harness Cloud">
```

Cache your Go module dependencies with [**Cache Intelligence**](/docs/continuous-integration/use-ci/caching-ci-data/cache-intelligence).

```yaml
    - stage:
        spec:
          caching:
            enabled: true
```

```mdx-code-block
</TabItem>

<TabItem value="Self-hosted">
```

See [Save and Restore Cache from S3](/docs/continuous-integration/use-ci/caching-ci-data/saving-cache/) or [Save and Restore Cache from GCS](/docs/continuous-integration/use-ci/caching-ci-data/save-cache-in-gcs).

```mdx-code-block
</TabItem>
</Tabs>
```

## Visualize test results

Test results can be [viewed](/docs/continuous-integration/use-ci/set-up-test-intelligence/viewing-tests/) in the **Tests** tab in your pipeline execution. Test results must be in JUnit XML format.

[go-junit-report](https://github.com/jstemmer/go-junit-report) can be used to output compatible JUnit XML reports.

> Modify your test step to generate the JUnit XML, and add the reports path.

Reviewer: ```suggestion
For your pipeline to produce test reports, you need to modify the **Run** step that runs your tests. Make sure the `command` generates JUnit XML reports and add the `reports` specification.
```

I am concerned that calling it "tests" might make people think there's a "tests" step or get confused with the "Run tests" step. Feel free to reword this as you like.

> Modify your test step to generate the JUnit XML, and add the reports path.

Reviewer: ```suggestion
For your pipeline to publish test reports to the **Tests** tab, you need to modify the **Run** step that runs your tests. Make sure the `command` generates JUnit XML reports and add the `reports` specification.
```

I am concerned that calling it "tests" might make people think there's a "tests" step or get confused with the "Run tests" step. Feel free to reword this as you like.

Author: This seems clear to me, thanks

                      cat report.out | $HOME/go/bin/go-junit-report -set-exit-code > report.xml
                    reports:
                      type: JUnit
                      spec:
                        paths:
                          - report.xml
```

```mdx-code-block
</TabItem>

<TabItem value="Self-hosted">
```

```yaml
              - step:
                  type: Run
                  identifier: test
                  name: Test
                  spec:
                    connectorRef: [your-docker-connector-ID] # replace with your connector ID
                    image: golang
                    command: |-
                      go install github.com/jstemmer/go-junit-report/v2@latest
                      go test -v ./... | tee report.out
                      cat report.out | $GOPATH/bin/go-junit-report -set-exit-code > report.xml
                    reports:
                      type: JUnit
                      spec:
                        paths:
                          - report.xml
```

```mdx-code-block
</TabItem>
</Tabs>
```

## Specify version

```mdx-code-block
<Tabs>
<TabItem value="Harness Cloud">
```

> Hosted Cloud runners come with Go pre-installed, see [Platforms and image specifications](/docs/continuous-integration/use-ci/set-up-build-infrastructure/use-harness-cloud-build-infrastructure#platforms-and-image-specifications) for all available tools and versions.

Reviewer: ```suggestion
Go is pre-installed on Hosted Cloud runners. For details about all available tools and versions, go to [Platforms and image specifications](/docs/continuous-integration/use-ci/set-up-build-infrastructure/use-harness-cloud-build-infrastructure#platforms-and-image-specifications).
```

                  name: Install Go
                  spec:
                    shell: Sh
                    # install version 1.20 of Go
                    command: |-
                      export GOPATH=$HOME/go
                      go install golang.org/dl/go1.20@latest
                      $GOPATH/bin/go1.20 download
```

</details>

<details>
<summary>Install multiple versions of Go</summary>

Add the [matrix looping strategy](/docs/platform/pipelines/looping-strategies-matrix-repeat-and-parallelism/) configuration to your stage.

```yaml
        strategy:
          matrix:
            # matrix strategy with Go versions 1.19 and 1.20
            goVersion:
              - "1.19"
              - "1.20"
```

Reference the matrix variable in your steps.

```yaml
              - step:
                  type: Run
                  identifier: installgo
                  name: Install Go
                  spec:
                    shell: Sh
                    command: |-
                      export GOPATH=$HOME/go
                      go install golang.org/dl/go<+matrix.goVersion>@latest
                      $GOPATH/bin/go<+matrix.goVersion> download

> ```

Reviewer: ~~~suggestion
1. Add the [matrix looping strategy](/docs/platform/pipelines/looping-strategies-matrix-repeat-and-parallelism/) configuration to your stage.

   ```yaml
           strategy:
             matrix:
               # matrix strategy with Go versions 1.19 and 1.20
               goVersion:
                 - "1.19"
                 - "1.20"
   ```

2. Reference the matrix variable in your steps.

   ```yaml
                 - step:
                     type: Run
                     identifier: installgo
                     name: Install Go
                     spec:
                       shell: Sh
                       command: |-
                         export GOPATH=$HOME/go
                         go install golang.org/dl/go<+matrix.goVersion>@latest
                         $GOPATH/bin/go<+matrix.goVersion> download
   ```
~~~

Since there are two steps separated by a code block, I think the numbering might help the second step stick out.

Author: Makes sense

</details>

```mdx-code-block
</TabItem>

<TabItem value="Self-hosted">
```

Specify the desired [Golang Docker image](https://hub.docker.com/_/golang) tag in your steps. There is no need for a separate install step when using Docker.

<details>
<summary>Build using a specific version of Go</summary>

```yaml
              - step:
                  type: Run
                  identifier: build
                  name: Build
                  spec:
                    connectorRef: [your-docker-connector-ID] # replace with your connector ID
                    # use version 1.20 of Go
                    image: golang:1.20
                    command: |-
                      go build
```

</details>

<details>
<summary>Build using multiple versions of Go</summary>

> Add the [matrix looping strategy](/docs/platform/pipelines/looping-strategies-matrix-repeat-and-parallelism/) configuration to your stage.

Reviewer: same suggestion here to use numbering

```yaml
        strategy:
          matrix:
            # matrix strategy with Go versions 1.19 and 1.20
            goVersion:
              - "1.19"
              - "1.20"
```

Reference the matrix variable in the `image` field of your steps.

```yaml
              - step:
                  type: Run
                  identifier: build
                  name: Build
                  spec:
                    connectorRef: [your-docker-connector-ID] # replace with your connector ID
                    image: golang:<+matrix.goVersion>
                    command: |-
                      go build
```

</details>

```mdx-code-block
</TabItem>
</Tabs>
```

## Full pipeline examples

> Full pipeline examples based on the steps above.

Reviewer: I don't see caching in these example so maybe clarify like:

These YAML examples define a single-stage pipeline with steps to install Go, build a Go app, and run tests.

Author: Caching is in the cloud examples, but not self-hosted

I can add some details around what is featured in these pipelines

```mdx-code-block
<Tabs>
<TabItem value="Harness Cloud">
```

Replace the bracketed values with corresponding values for your [code repo connector](/docs/continuous-integration/use-ci/codebase-configuration/create-and-configure-a-codebase/#code-repo-connectors) and repository name.

Depending on your project and organization, you may also need to replace `projectIdentifier` and `orgIdentifier`.

<details>

> <summary>Use a specific Go version</summary>

Reviewer: ```suggestion
<summary>Pipeline with one specific Go version</summary>
```

                    shell: Sh
                    command: |-
                      export GOPATH=$HOME/go
                      $GOPATH/bin/go1.20 build
              - step:
                  type: Run
                  identifier: test
                  name: Test
                  spec:
                    shell: Sh
                    command: |-
                      export GOPATH=$HOME/go
                      $GOPATH/bin/go1.20 install github.com/jstemmer/go-junit-report/v2@latest
                      $GOPATH/bin/go1.20 test -v | tee report.out
                      cat report.out | $GOPATH/bin/go-junit-report -set-exit-code > report.xml
                    reports:
                      type: JUnit
                      spec:
                        paths:
                          - report.xml
          platform:
            os: Linux
            arch: Amd64
          runtime:
            type: Cloud
            spec: {}
  properties:
    ci:
      codebase:
        connectorRef: [your-code-repo-connector-ID] # replace with your connector ID
        repoName: [your-repository-name] # replace with your repository name
        build: <+input>
```

</details>

<details>

> <summary>Use multiple Go versions</summary>

Reviewer: ```suggestion
<summary>Pipeline with multiple Go versions</summary>
```

                    reports:
                      type: JUnit
                      spec:
                        paths:
                          - report_*.xml
          platform:
            os: Linux
            arch: Amd64
          runtime:
            type: Cloud
            spec: {}
  properties:
    ci:
      codebase:
        connectorRef: [your-code-repo-connector-ID] # replace with your connector ID
        repoName: [your-repository-name] # replace with your repository name
        build: <+input>
```

</details>

```mdx-code-block
</TabItem>

<TabItem value="Self-hosted">
```

Replace the bracketed values with corresponding values for your [code repo connector](/docs/continuous-integration/use-ci/codebase-configuration/create-and-configure-a-codebase/#code-repo-connectors), [kubernetes cluster connector](/docs/platform/Connectors/Cloud-providers/add-a-kubernetes-cluster-connector), kubernetes namespace, and repository name.

Depending on your project and organization, you may also need to replace `projectIdentifier` and `orgIdentifier`.

<details>

> <summary>Use a specific Go version</summary>

Reviewer: ```suggestion
<summary>Pipeline with one specific Go version</summary>
```

                  identifier: test
                  name: Test
                  spec:
                    connectorRef: [your-docker-connector-ID] # replace with your connector ID
                    image: golang:1.20
                    command: |-
                      go install github.com/jstemmer/go-junit-report/v2@latest
                      go test -v ./... | tee report.out
                      cat report.out | go-junit-report -set-exit-code > report.xml
                    reports:
                      type: JUnit
                      spec:
                        paths:
                          - report.xml
          infrastructure:
            spec:
              connectorRef: [your-kube-connector-ID] # replace with your connector ID
              namespace: [your-kube-namespace] # replace with your namespace
            type: KubernetesDirect
          platform:
            arch: Amd64
            os: Linux
  properties:
    ci:
      codebase:
        connectorRef: [your-code-repo-connector-ID] # replace with your connector ID
        repoName: [your-repository-name] # replace with your repository name
        build: <+input>
```

</details>

<details>

> <summary>Use multiple Go versions</summary>

Reviewer: ```suggestion
<summary>Pipeline with multiple Go versions</summary>
```

                    image: golang:<+matrix.goVersion>
                    command: |-
                      go install github.com/jstemmer/go-junit-report/v2@latest
                      go test -v ./... | tee report_<+matrix.goVersion>.out
                      cat report_<+matrix.goVersion>.out | $GOPATH/bin/go-junit-report -set-exit-code > report_<+matrix.goVersion>.xml
                    reports:
                      type: JUnit
                      spec:
                        paths:
                          - report_*.xml
          infrastructure:
            spec:
              connectorRef: [your-kube-connector-ID] # replace with your connector ID
              namespace: [your-kube-namespace] # replace with your namespace
            type: KubernetesDirect
          platform:
            arch: Amd64
            os: Linux
  properties:
    ci:
      codebase:
        connectorRef: [your-code-repo-connector-ID] # replace with your connector ID
        repoName: [your-repository-name] # replace with your repository name
        build: <+input>
```

</details>

```mdx-code-block
</TabItem>
</Tabs>
```

## Next steps

> Create [triggers](https://developer.harness.io/docs/category/triggers) to execute your pipeline.
> 
> Learn how to [build and upload artifacts](/docs/category/build-and-upload-artifacts), or how to [build and push an image to a Docker registry](/docs/continuous-integration/use-ci/build-and-upload-artifacts/build-and-push-to-docker-hub-step-settings/).

Reviewer: ```suggestion
Now that you have created a pipeline that builds and tests a Go app, you could:

* Create [triggers](https://developer.harness.io/docs/category/triggers) to automatically run your pipeline.
* Add steps to [build and upload artifacts](/docs/category/build-and-upload-artifacts).
* Add a step to [build and push an image to a Docker registry](/docs/continuous-integration/use-ci/build-and-upload-artifacts/build-and-push-to-docker-hub-step-settings/).
```
